yt_dlp/extractor/zdf.py

```diff
@@ -132,16 +130,31 @@ def _extract_ptmd(self, ptmd_urls, video_id, api_token=None, aspect_ratio=None):
                                 'format_id': join_nonempty('http', stream.get('type')),
                                 'tbr': int_or_none(self._search_regex(r'_(\d+)k_', format_url, 'tbr', default=None)),
                             }]
+                        else:
+                            self.write_debug(f'Unsupported extension {ext} ({format_url}).')
+                            fmts = []
+
                         f_class = variant.get('class')
                         for f in fmts:
+                            f_lang = ISO639Utils.short2long(
+                                (f.get('language') or variant.get('language') or '').lower())
+                            has_video = (f.get('vcodec') or 'none') != 'none'
```

Comment:
wouldn't `has_audio` be more to the point? I don't like defaulting to `'none'`
```suggestion
                            has_audio = f.get('acodec') != 'none'
```

Reply:
I kindof need `is_audio_only`. Otherwise pre-merged HTTP formats will miss their format_note and pre-merged `ad` formats are treated as normal formats.

```diff
diff --git a/yt_dlp/extractor/zdf.py b/yt_dlp/extractor/zdf.py
index 1da807292..49ba9846b 100644
--- a/yt_dlp/extractor/zdf.py
+++ b/yt_dlp/extractor/zdf.py
@@ -138,19 +138,19 @@ def _extract_ptmd(self, ptmd_urls, video_id, api_token=None, aspect_ratio=None):
                         for f in fmts:
                             f_lang = ISO639Utils.short2long(
                                 (f.get('language') or variant.get('language') or '').lower())
-                            has_video = (f.get('vcodec') or 'none') != 'none'
+                            has_audio = f.get('acodec') != 'none'
                             formats.append({
                                 **f,
                                 'format_id': join_nonempty(f['format_id'], is_dgs and 'dgs'),
                                 'format_note': join_nonempty(
-                                    has_video and f_class,
+                                    not has_audio and f_class,
                                     is_dgs and 'German Sign Language',
                                     f.get('format_note'), delim=', '),
                                 'preference': -2 if is_dgs else -1,
                                 'language': f_lang,
                                 'language_preference': (
-                                    -10 if ((not has_video and f.get('format_note') == 'Audiodeskription')
-                                            or (has_video and f_class == 'ad'))
+                                    -10 if ((has_audio and f.get('format_note') == 'Audiodeskription')
+                                            or (not has_audio and f_class == 'ad'))
                                     else 10 if f_lang == 'deu' and f_class == 'main'
                                     else 5 if f_lang == 'deu'
                                     else 1 if f_class == 'main'

```

<details><summary>Resulting formats</summary> 

```
$ yt-dlp -qF https://www.zdf.de/video/serien/the-tourist-110/iren-sind-menschlich-100
ID                                 EXT  RESOLUTION FPS │   FILESIZE   TBR PROTO │ VCODEC          VBR ACODEC     MORE INFO
───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
hls-audio0-Audiodeskription-0      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-1      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-2      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-3      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-4      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-5      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-6      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-7      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-8      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-9      mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-10     mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Audiodeskription-11     mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] Audiodeskription
hls-audio0-Originalton_englisch-0  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-1  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-2  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-3  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-4  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-5  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-6  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-7  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-8  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-9  mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-10 mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-Originalton_englisch-11 mp4  audio only     │                  m3u8  │ audio only          unknown    [eng] Originalton englisch
hls-audio0-TV_Ton-0                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-1                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-2                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-3                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-4                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-5                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-6                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-7                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-8                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-9                mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-10               mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-audio0-TV_Ton-11               mp4  audio only     │                  m3u8  │ audio only          unknown    [deu] TV Ton
hls-321-0                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] ad
hls-321-1                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] ad
hls-321-2                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] ad
hls-321-3                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] ad
hls-321-4                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] ad
hls-321-5                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] ad
hls-321-6                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] ad
hls-321-7                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] ad
hls-321-8                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] ad
hls-559-0                          mp4  640x360     25 │ ~237.09MiB  560k m3u8  │ avc1.4d401f    560k video only [deu] ad
hls-559-1                          mp4  640x360     25 │ ~237.09MiB  560k m3u8  │ avc1.4d401f    560k video only [deu] ad
hls-559-2                          mp4  640x360     25 │ ~237.09MiB  560k m3u8  │ avc1.4d401f    560k video only [deu] ad
hls-1223-0                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] ad
hls-1223-1                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] ad
hls-1223-2                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] ad
hls-1223-3                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] ad
hls-1223-4                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] ad
hls-2624-0                         mp4  1280x720    50 │ ~  1.09GiB 2624k m3u8  │ avc1.640028   2624k video only [deu] ad
hls-2624-1                         mp4  1280x720    50 │ ~  1.09GiB 2624k m3u8  │ avc1.640028   2624k video only [deu] ad
hls-5332-0                         mp4  1920x1080   50 │ ~  2.21GiB 5333k m3u8  │ avc1.64002a   5333k video only [deu] ad
hls-5332-1                         mp4  1920x1080   50 │ ~  2.21GiB 5333k m3u8  │ avc1.64002a   5333k video only [deu] ad
http-h264_aac_mp4_http_na_na-0     mp4  480x270        │ ≈215.10MiB  508k https │ avc1.4d401f         mp4a.40.2  [eng]
http-vp9_opus_webm_http_na_na-0    webm 480x270        │ ≈155.82MiB  368k https │ vp9                 opus       [eng]
hls-321-9                          mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [eng] ot
hls-321-10                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [eng] ot
hls-321-11                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [eng] ot
hls-321-12                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [eng] ot
hls-321-13                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [eng] ot
hls-321-14                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [eng] ot
hls-321-15                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [eng] ot
hls-321-16                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [eng] ot
hls-321-17                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [eng] ot
http-h264_aac_mp4_http_na_na-1     mp4  640x360        │ ≈342.13MiB  808k https │ avc1.4d401f         mp4a.40.2  [eng]
http-h264_aac_mp4_http_na_na-2     mp4  640x360        │ ≈342.13MiB  808k https │ avc1.4d401f         mp4a.40.2  [eng]
hls-559-3                          mp4  640x360     25 │ ~237.09MiB  560k m3u8  │ avc1.4d401f    560k video only [eng] ot
hls-559-4                          mp4  640x360     25 │ ~237.09MiB  560k m3u8  │ avc1.4d401f    560k video only [eng] ot
hls-559-5                          mp4  640x360     25 │ ~237.09MiB  560k m3u8  │ avc1.4d401f    560k video only [eng] ot
http-h264_aac_mp4_http_na_na-3     mp4  960x540        │ ≈689.35MiB 1628k https │ avc1.4d401f         mp4a.40.2  [eng]
http-h264_aac_mp4_http_na_na-4     mp4  960x540        │ ≈689.35MiB 1628k https │ avc1.4d401f         mp4a.40.2  [eng]
http-vp9_opus_webm_http_na_na-1    webm 960x540        │ ≈477.63MiB 1128k https │ vp9                 opus       [eng]
hls-1223-5                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [eng] ot
hls-1223-6                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [eng] ot
hls-1223-7                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [eng] ot
hls-1223-8                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [eng] ot
hls-1223-9                         mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [eng] ot
http-h264_aac_mp4_http_na_na-5     mp4  1280x720       │ ≈  1.38GiB 3328k https │ avc1.640028         mp4a.40.2  [eng]
http-vp9_opus_webm_http_na_na-2    webm 1280x720       │ ≈901.06MiB 2128k https │ vp9                 opus       [eng]
hls-2624-2                         mp4  1280x720    50 │ ~  1.09GiB 2624k m3u8  │ avc1.640028   2624k video only [eng] ot
hls-2624-3                         mp4  1280x720    50 │ ~  1.09GiB 2624k m3u8  │ avc1.640028   2624k video only [eng] ot
http-h264_aac_mp4_http_na_na-6     mp4  1920x1080      │ ≈  2.74GiB 6628k https │ avc1.64002a         mp4a.40.2  [eng]
http-vp9_opus_webm_http_na_na-3    webm 1920x1080      │ ≈  1.79GiB 4328k https │ vp9                 opus       [eng]
hls-5332-2                         mp4  1920x1080   50 │ ~  2.21GiB 5333k m3u8  │ avc1.64002a   5333k video only [eng] ot
hls-5332-3                         mp4  1920x1080   50 │ ~  2.21GiB 5333k m3u8  │ avc1.64002a   5333k video only [eng] ot
http-h265_aac_mp4_http_na_na-0     mp4  3840x2160      │ ≈  1.94GiB 4692k https │ hvc1.1.7.L120       mp4a.40.2  [eng]
http-h264_aac_mp4_http_na_na-7     mp4  480x270        │ ≈215.10MiB  508k https │ avc1.4d401f         mp4a.40.2  [deu]
http-vp9_opus_webm_http_na_na-4    webm 480x270        │ ≈155.82MiB  368k https │ vp9                 opus       [deu]
http-h264_aac_mp4_http_na_na-8     mp4  640x360        │ ≈342.13MiB  808k https │ avc1.4d401f         mp4a.40.2  [deu]
http-h264_aac_mp4_http_na_na-9     mp4  640x360        │ ≈342.13MiB  808k https │ avc1.4d401f         mp4a.40.2  [deu]
http-h264_aac_mp4_http_na_na-10    mp4  960x540        │ ≈689.35MiB 1628k https │ avc1.4d401f         mp4a.40.2  [deu]
http-h264_aac_mp4_http_na_na-11    mp4  960x540        │ ≈689.35MiB 1628k https │ avc1.4d401f         mp4a.40.2  [deu]
http-vp9_opus_webm_http_na_na-5    webm 960x540        │ ≈477.63MiB 1128k https │ vp9                 opus       [deu]
http-h264_aac_mp4_http_na_na-12    mp4  1280x720       │ ≈  1.38GiB 3328k https │ avc1.640028         mp4a.40.2  [deu]
http-vp9_opus_webm_http_na_na-6    webm 1280x720       │ ≈901.06MiB 2128k https │ vp9                 opus       [deu]
http-h264_aac_mp4_http_na_na-13    mp4  1920x1080      │ ≈  2.74GiB 6628k https │ avc1.64002a         mp4a.40.2  [deu]
http-vp9_opus_webm_http_na_na-7    webm 1920x1080      │ ≈  1.79GiB 4328k https │ vp9                 opus       [deu]
http-h265_aac_mp4_http_na_na-1     mp4  3840x2160      │ ≈  1.94GiB 4692k https │ hvc1.1.7.L120       mp4a.40.2  [deu]
http-h264_aac_mp4_http_na_na-14    mp4  480x270        │ ≈215.10MiB  508k https │ avc1.4d401f         mp4a.40.2  [deu]
http-vp9_opus_webm_http_na_na-8    webm 480x270        │ ≈155.82MiB  368k https │ vp9                 opus       [deu]
hls-321-18                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] main
hls-321-19                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] main
hls-321-20                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] main
hls-321-21                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] main
hls-321-22                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] main
hls-321-23                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] main
hls-321-24                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] main
hls-321-25                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] main
hls-321-26                         mp4  480x270     25 │ ~136.05MiB  321k m3u8  │ avc1.4d401f    321k video only [deu] main
http-h264_aac_mp4_http_na_na-15    mp4  640x360        │ ≈342.13MiB  808k https │ avc1.4d401f         mp4a.40.2  [deu]
http-h264_aac_mp4_http_na_na-16    mp4  640x360        │ ≈342.13MiB  808k https │ avc1.4d401f         mp4a.40.2  [deu]
hls-559-6                          mp4  640x360     25 │ ~237.09MiB  560k m3u8  │ avc1.4d401f    560k video only [deu] main
hls-559-7                          mp4  640x360     25 │ ~237.09MiB  560k m3u8  │ avc1.4d401f    560k video only [deu] main
hls-559-8                          mp4  640x360     25 │ ~237.09MiB  560k m3u8  │ avc1.4d401f    560k video only [deu] main
http-h264_aac_mp4_http_na_na-17    mp4  960x540        │ ≈689.35MiB 1628k https │ avc1.4d401f         mp4a.40.2  [deu]
http-h264_aac_mp4_http_na_na-18    mp4  960x540        │ ≈689.35MiB 1628k https │ avc1.4d401f         mp4a.40.2  [deu]
http-vp9_opus_webm_http_na_na-9    webm 960x540        │ ≈477.63MiB 1128k https │ vp9                 opus       [deu]
hls-1223-10                        mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] main
hls-1223-11                        mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] main
hls-1223-12                        mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] main
hls-1223-13                        mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] main
hls-1223-14                        mp4  960x540     25 │ ~518.19MiB 1224k m3u8  │ avc1.4d401f   1224k video only [deu] main
http-h264_aac_mp4_http_na_na-19    mp4  1280x720       │ ≈  1.38GiB 3328k https │ avc1.640028         mp4a.40.2  [deu]
http-vp9_opus_webm_http_na_na-10   webm 1280x720       │ ≈901.06MiB 2128k https │ vp9                 opus       [deu]
hls-2624-4                         mp4  1280x720    50 │ ~  1.09GiB 2624k m3u8  │ avc1.640028   2624k video only [deu] main
hls-2624-5                         mp4  1280x720    50 │ ~  1.09GiB 2624k m3u8  │ avc1.640028   2624k video only [deu] main
http-h264_aac_mp4_http_na_na-20    mp4  1920x1080      │ ≈  2.74GiB 6628k https │ avc1.64002a         mp4a.40.2  [deu]
http-vp9_opus_webm_http_na_na-11   webm 1920x1080      │ ≈  1.79GiB 4328k https │ vp9                 opus       [deu]
hls-5332-4                         mp4  1920x1080   50 │ ~  2.21GiB 5333k m3u8  │ avc1.64002a   5333k video only [deu] main
hls-5332-5                         mp4  1920x1080   50 │ ~  2.21GiB 5333k m3u8  │ avc1.64002a   5333k video only [deu] main
http-h265_aac_mp4_http_na_na-2     mp4  3840x2160      │ ≈  1.94GiB 4692k https │ hvc1.1.7.L120       mp4a.40.2  [deu]
```
</details>

I could do `is_audio_only = f.get('acodec') != 'none' and f.get('vcodec') == 'none'` though, which behaves mostly the same as my current version, but slightly better for https://www.zdf.de/play/video/funk-collection-funk-1073-108/funk-fuehrer-joon-kims-eiserne-faust---kow-10-100

Reply:
You should be able to use `f.get('vcodec') == 'none'` alone to determine if it `is_audio_only`

Reply:
I didn't do that initially because the video I linked has a pre-merged HTTP format that has unknown `acodec`/`vcodec`. But I guess this works out because that can only happen for the pre-merged format since the HLS formats always provide one.

Reply:
if the video codec is unknown but a video track is present in the format, `f.get('vcodec')` will return `None`. the string `'none'` is specifically for when yt-dlp knows there is NO codec

(same applies to `'acodec'`)

Reply:
Yeah, that is how I meant it in my comment as well. The scenario I was trying to guard against with the extra `or 'none'` seemed sensible to me at first, but isn't actually.